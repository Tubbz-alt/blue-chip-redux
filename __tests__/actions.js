import jsonApiPayload
  from "../__testHelpers__/fixtures/checklistsJsonApiResponse";
import graphQlPayload
  from "../__testHelpers__/fixtures/checklistsGraphQlResponse";
import normalizedJsonApiTasksPayload
  from "../__testHelpers__/fixtures/normalizedJsonApiTasksPayload";
import normalizedJsonApiChecklistsPayload
  from "../__testHelpers__/fixtures/normalizedJsonApiChecklistsPayload";
import normalizedGraphQlTaskPayload
  from "../__testHelpers__/fixtures/normalizedGraphQlTasksPayload";
import normalizedGraphQlChecklistPayload
  from "../__testHelpers__/fixtures/normalizedGraphQlChecklistsPayload";

import {Actions} from "../../blue-chip";
import reduxAdapter from "../";
const dispatch = jest.fn().mockName("dispatch");
const actions = Actions.config({
  adapter: reduxAdapter,
  mutator: dispatch
});

describe("Redux Actions", () => {
  describe("updateResources", () => {
    describe("JsonApi", () => {
      test("dispatches UPDATE_RESOURCES for each ", () => {
        const tasksMergeResourcesAction = {
          resourceType: "tasks",
          resourcesById: normalizedJsonApiTasksPayload,
          type: "UPDATE_RESOURCES"
        };

        const checklistsMergeResourcesAction = {
          resourceType: "checklists",
          resourcesById: normalizedJsonApiChecklistsPayload,
          type: "UPDATE_RESOURCES"
        };

        actions.updateResources(jsonApiPayload);
        expect(dispatch).toBeCalledWith(tasksMergeResourcesAction);
        expect(dispatch).toBeCalledWith(checklistsMergeResourcesAction);
      });
    });

    describe("GraphQl", () => {
      test("dispatches UPDATE_RESOURCES for each ", () => {
        const tasksMergeResourcesAction = {
          resourceType: "tasks",
          resourcesById: normalizedGraphQlTaskPayload,
          type: "UPDATE_RESOURCES"
        };

        const checklistsMergeResourcesAction = {
          resourceType: "checklists",
          resourcesById: normalizedGraphQlChecklistPayload,
          type: "UPDATE_RESOURCES"
        };

        actions.updateResources(graphQlPayload);
        expect(dispatch).toBeCalledWith(tasksMergeResourcesAction);
        expect(dispatch).toBeCalledWith(checklistsMergeResourcesAction);
      });
    });
  });

  describe("updateResource", () => {
    test("dispatches update action for a single resource", () => {
      const checklist = {
        id: 1,
        type: "checklists",
        attributes: {name: "Onboarding Rest"},
        links: {self: "http://example.com/checklists/1"},
        relationships: {
          tasks: {
            data: [{id: 1, type: "tasks"}, {id: 2, type: "tasks"}]
          }
        }
      };

      const updateAction = {
        type: "ADD_OR_REPLACE_RESOURCE_BY_ID",
        resourceType: checklist.type,
        id: checklist.id,
        attributes: checklist.attributes,
        links: checklist.links,
        relationships: checklist.relationships
      };

      actions.updateResource(checklist);
      expect(dispatch).toBeCalledWith(updateAction);
    });
  });

  describe("removeResources", () => {
    test("dispatches remove resource action", () => {
      const checklists = [
        {
          id: 1,
          type: "checklists"
        },
        {
          id: 2,
          type: "checklists"
        }
      ];

      const removeResourcesAction = {
        type: "REMOVE_RESOURCES_BY_ID",
        resources: checklists
      };

      actions.removeResources(checklists);
      expect(dispatch).toBeCalledWith(removeResourcesAction);
    });
  });

  describe("removeResource", () => {
    test("dispatches remove resource action", () => {
      const checklist = {
        id: 1,
        type: "checklists"
      };

      const deleteAction = {
        type: "REMOVE_RESOURCE_BY_ID",
        resourceType: checklist.type,
        id: checklist.id
      };

      actions.removeResource(checklist);
      expect(dispatch).toBeCalledWith(deleteAction);
    });
  });

  describe("clearResources", () => {
    test("clears the store for the provided resources", () => {
      const resourceTypes = ["checklists", "tasks"];

      const clearResourcesAction = {
        type: "CLEAR_RESOURCES",
        resourceTypes
      };

      actions.clearResources(resourceTypes, dispatch);
      expect(dispatch).toBeCalledWith(clearResourcesAction);
    });
  });
});
