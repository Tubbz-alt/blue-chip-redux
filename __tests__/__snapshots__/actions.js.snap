// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Redux Actions updateResources GraphQl dispatches UPDATE_RESOURCES for each  1`] = `
[MockFunction dispatch] {
  "calls": Array [
    Array [
      Object {
        "index": Array [
          1,
          2,
          3,
        ],
        "resourcesByType": Object {
          "checklists": Object {
            "1": Object {
              "attributes": Object {
                "name": "Onboarding Rest",
              },
              "id": 1,
              "links": Object {
                "self": "http://example.com/checklists/1",
              },
              "relationships": Object {
                "tasks": Object {
                  "data": Array [
                    Object {
                      "id": 1,
                      "type": "tasks",
                    },
                    Object {
                      "id": 2,
                      "type": "tasks",
                    },
                  ],
                  "links": Object {
                    "related": "http://example.com/checklists/1/tasks",
                    "self": "http://example.com/checklists/1/relationships/tasks",
                  },
                },
              },
              "type": "checklists",
            },
            "2": Object {
              "attributes": Object {
                "name": "Project Audit Rest",
              },
              "id": 2,
              "links": Object {
                "self": "http://example.com/checklists/2",
              },
              "relationships": Object {
                "tasks": Object {
                  "data": Array [
                    Object {
                      "id": 3,
                      "type": "tasks",
                    },
                    Object {
                      "id": 4,
                      "type": "tasks",
                    },
                  ],
                  "links": Object {
                    "related": "http://example.com/checklists/2/tasks",
                    "self": "http://example.com/checklists/2/relationships/tasks",
                  },
                },
              },
              "type": "checklists",
            },
            "3": Object {
              "attributes": Object {
                "name": "QA Checklist Rest",
              },
              "id": 3,
              "links": Object {
                "self": "http://example.com/checklists/3",
              },
              "relationships": Object {
                "tasks": Object {
                  "data": Array [
                    Object {
                      "id": 5,
                      "type": "tasks",
                    },
                    Object {
                      "id": 6,
                      "type": "tasks",
                    },
                  ],
                  "links": Object {
                    "related": "http://example.com/checklists/3/tasks",
                    "self": "http://example.com/checklists/3/relationships/tasks",
                  },
                },
              },
              "type": "checklists",
            },
          },
          "tasks": Object {
            "1": Object {
              "attributes": Object {
                "description": "Onboarding REST 1",
              },
              "id": 1,
              "links": Object {
                "self": "http://example.com/tasks/1",
              },
              "type": "tasks",
            },
            "2": Object {
              "attributes": Object {
                "description": "Onboarding REST 2",
              },
              "id": 2,
              "links": Object {
                "self": "http://example.com/tasks/2",
              },
              "type": "tasks",
            },
            "3": Object {
              "attributes": Object {
                "description": "Project Audit Rest 3",
              },
              "id": 3,
              "links": Object {
                "self": "http://example.com/tasks/3",
              },
              "type": "tasks",
            },
            "4": Object {
              "attributes": Object {
                "description": "Project Audit Rest 4",
              },
              "id": 4,
              "links": Object {
                "self": "http://example.com/tasks/4",
              },
              "type": "tasks",
            },
            "5": Object {
              "attributes": Object {
                "description": "QA Checklist Rest 5",
              },
              "id": 5,
              "links": Object {
                "self": "http://example.com/tasks/5",
              },
              "type": "tasks",
            },
            "6": Object {
              "attributes": Object {
                "description": "QA Checklist Rest 6",
              },
              "id": 6,
              "links": Object {
                "self": "http://example.com/tasks/6",
              },
              "type": "tasks",
            },
          },
        },
        "type": "UPDATE_RESOURCES",
      },
    ],
    Array [
      Object {
        "index": Array [],
        "resourcesByType": Object {
          "checklists": Object {
            "4": Object {
              "id": 4,
              "name": "Onboarding GraphQL",
              "tasks": Array [
                7,
                8,
              ],
            },
            "5": Object {
              "id": 5,
              "name": "Project Audit GraphQL",
              "tasks": Array [
                9,
                10,
              ],
            },
            "6": Object {
              "id": 6,
              "name": "QA Checklist GraphQL",
              "tasks": Array [
                11,
                12,
              ],
            },
          },
          "tasks": Object {
            "10": Object {
              "description": "Project Audit Graph 10",
              "id": 10,
            },
            "11": Object {
              "description": "QA Checklist Graph 11",
              "id": 11,
            },
            "12": Object {
              "description": "QA Checklist Graph 12",
              "id": 12,
            },
            "7": Object {
              "description": "Onboarding Graph 7",
              "id": 7,
            },
            "8": Object {
              "description": "Onboarding Graph 8",
              "id": 8,
            },
            "9": Object {
              "description": "Project Audit Graph 9",
              "id": 9,
            },
          },
        },
        "type": "UPDATE_RESOURCES",
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;

exports[`Redux Actions updateResources JsonApi dispatches UPDATE_RESOURCES for each  1`] = `
[MockFunction dispatch] {
  "calls": Array [
    Array [
      Object {
        "index": Array [
          1,
          2,
          3,
        ],
        "resourcesByType": Object {
          "checklists": Object {
            "1": Object {
              "attributes": Object {
                "name": "Onboarding Rest",
              },
              "id": 1,
              "links": Object {
                "self": "http://example.com/checklists/1",
              },
              "relationships": Object {
                "tasks": Object {
                  "data": Array [
                    Object {
                      "id": 1,
                      "type": "tasks",
                    },
                    Object {
                      "id": 2,
                      "type": "tasks",
                    },
                  ],
                  "links": Object {
                    "related": "http://example.com/checklists/1/tasks",
                    "self": "http://example.com/checklists/1/relationships/tasks",
                  },
                },
              },
              "type": "checklists",
            },
            "2": Object {
              "attributes": Object {
                "name": "Project Audit Rest",
              },
              "id": 2,
              "links": Object {
                "self": "http://example.com/checklists/2",
              },
              "relationships": Object {
                "tasks": Object {
                  "data": Array [
                    Object {
                      "id": 3,
                      "type": "tasks",
                    },
                    Object {
                      "id": 4,
                      "type": "tasks",
                    },
                  ],
                  "links": Object {
                    "related": "http://example.com/checklists/2/tasks",
                    "self": "http://example.com/checklists/2/relationships/tasks",
                  },
                },
              },
              "type": "checklists",
            },
            "3": Object {
              "attributes": Object {
                "name": "QA Checklist Rest",
              },
              "id": 3,
              "links": Object {
                "self": "http://example.com/checklists/3",
              },
              "relationships": Object {
                "tasks": Object {
                  "data": Array [
                    Object {
                      "id": 5,
                      "type": "tasks",
                    },
                    Object {
                      "id": 6,
                      "type": "tasks",
                    },
                  ],
                  "links": Object {
                    "related": "http://example.com/checklists/3/tasks",
                    "self": "http://example.com/checklists/3/relationships/tasks",
                  },
                },
              },
              "type": "checklists",
            },
          },
          "tasks": Object {
            "1": Object {
              "attributes": Object {
                "description": "Onboarding REST 1",
              },
              "id": 1,
              "links": Object {
                "self": "http://example.com/tasks/1",
              },
              "type": "tasks",
            },
            "2": Object {
              "attributes": Object {
                "description": "Onboarding REST 2",
              },
              "id": 2,
              "links": Object {
                "self": "http://example.com/tasks/2",
              },
              "type": "tasks",
            },
            "3": Object {
              "attributes": Object {
                "description": "Project Audit Rest 3",
              },
              "id": 3,
              "links": Object {
                "self": "http://example.com/tasks/3",
              },
              "type": "tasks",
            },
            "4": Object {
              "attributes": Object {
                "description": "Project Audit Rest 4",
              },
              "id": 4,
              "links": Object {
                "self": "http://example.com/tasks/4",
              },
              "type": "tasks",
            },
            "5": Object {
              "attributes": Object {
                "description": "QA Checklist Rest 5",
              },
              "id": 5,
              "links": Object {
                "self": "http://example.com/tasks/5",
              },
              "type": "tasks",
            },
            "6": Object {
              "attributes": Object {
                "description": "QA Checklist Rest 6",
              },
              "id": 6,
              "links": Object {
                "self": "http://example.com/tasks/6",
              },
              "type": "tasks",
            },
          },
        },
        "type": "UPDATE_RESOURCES",
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;
